File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> chars;
        int maxe = 0;
        int l = 0;

        for(int r= 0;r < s.length();++r){
            while(chars.find(s[r]) != chars.end()){
                chars.erase(s[l]);
                ++l;
            }
            chars.insert(s[r]);
            maxe = max(maxe, r-l+1);
        }
        return maxe;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l,r = 0,0
        length = len(s) 
        if not length:
            return 0
        seen = set()
        maxi = 1
        while r != length:
            if s[r] not in seen:
                seen.add(s[r])
                r += 1
            else:
                maxi = max(maxi,len(seen))
                while s[r] in seen:
                    # print(r,l)
                    seen.remove(s[l])
                    l +=1

        maxi = max(maxi,len(seen))
        return maxi
                

